File: 0841. Keys and Rooms.cs
public class Solution
{
    public bool CanVisitAllRooms(IList<IList<int>> rooms)
    {
        var isUsed = new HashSet<int>();
        Dfs(0, rooms, isUsed);

        return isUsed.Count == rooms.Count();
    }

    public void Dfs(int v, IList<IList<int>> graph, HashSet<int> isUsed)
    {
        isUsed.Add(v);

        foreach (var u in graph[v])
        {
            if (!isUsed.Contains(u))
            {
                Dfs(u, graph, isUsed);
            }
        }
    }
}

File: 0841. Keys and Rooms.cpp
class Solution
{
public:
    bool canVisitAllRooms(vector<vector<int>>& rooms)
    {
        vector<bool> visited(rooms.size(), false);

        dfs(0, rooms, visited);
        
        for (int i = 0; i < rooms.size(); ++i)
        {
            if (not visited[i])
            {
                return false;
            }
        }
        
        return true;
    }
    
    void dfs(int v, vector<vector<int>>& graph, vector<bool>& visited)
    {
        visited[v] = true;
        
        for (int u : graph[v])
        {
            if (not visited[u])
            {
                dfs(u, graph, visited);
            }   
        }
    }
};